056-zk-kraft-migration.md

```diff
@@ -0,0 +1,451 @@
+* `migration`: indicates the user's intention to move the migration process forward with brokers configured with the ZooKeeper migration enabled, and connected to both ZooKeeper nodes and the KRaft controllers, so that the Apache Kafka migration component can run the actual metadata migration ending with the "dual write" mode where metadata are both written to ZooKeeper and KRaft. Brokers are also connected to both ZooKeeper nodes and KRaft controllers.
+* `post-migration`: indicates the user's intention to finalize the migration by disabling the ZooKeeper migration on controllers and brokers, rolling the nodes, and finally having a full KRaft-based cluster with ZooKeeper ensemble still running but not used by brokers anymore.
+
+Another "input" is represented by the Kafka migration metrics checked by the operator during the actual metadata migration from ZooKeeper to KRaft.
+
+The state machine is described in the following diagram:
+
+![Migration Finite State Machine](./images/056-zk-kraft-migration-fsm.png)
+
+The following sections explain the states and transitions in details.
+The "Phase"(s) reflect what is described in the official [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) guide in the Apache Kafka documentation but with the right degree of automation provided by the operator.
+Also, the `PRE_MIGRATION`, `MIGRATION` and `POST_MIGRATION` dotted boxes correspond to the `ZkMigrationState` [enum values](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/migration/ZkMigrationState.java#L32) from the Apache Kafka migration component.
+
+In order to make the process clearer as much as possible, each phase is described by:
+
+* _Preconditions_: conditions that should be met (i.e. current state in the migration FSM) in order to have the operator taking actions on a trigger.
+* _Trigger_: what makes the operator actions to run only with verified preconditions.
+* _Actions_: what the operator does when preconditions are met (i.e. nodes configuration, rolling, ...) together with things that could be observable during actions (i.e. Kafka metrics changes).
+* _Status changes_: how the FSM changes status when the actions are completed.
+
+#### Already running KRaft-based clusters
+
+**_Preconditions_**
+
+The Apache Kafka cluster is running in KRaft mode, so the `strimzi.io/kraft: enabled` annotation is already applied on the `Kafka` custom resource.
+The FSM is in the `None` state (as per `Kafka.status.migrationState` field).
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: pre-migration` annotation on the `Kafka` custom resource.
```

Comment:
How do you track the pre-condition / condition? Clearly, `strimzi.io/kraft` cannot have two states at the same time. So how do you evaluate this?

Reply:
It depends on the state of the FSM. The transition is defined by the current state + the input which is the value changed on the `strimzi.io/kraft` annotation.
In this specific case, on first reconcile with the new operator version, the presence of `strimzi.io/kraft: enabled` will move the state into `None`. From there the FSM transitions are well defined (see the diagram).

Reply:
So, I assume the precondition should not mention the annotation in that case as it is confusing when you talk about checking something that will not be there anymore. the `None` and `Nomigration` states seem also quite confusing. Should they be `ZooKeeper` and `KRaft` instead to make it more understandable?

Reply:
> So, I assume the precondition should not mention the annotation in that case as it is confusing when you talk about checking something that will not be there anymore.

Why? The precondition highlights what should be the value of the annotation which will change due to the trigger, but the annotation value is part of the initial state of the FSM. 

> the None and Nomigration states seem also quite confusing. Should they be ZooKeeper and KRaft instead to make it more understandable?

Yeah I think these names could be have more sense. Only thing is about the `FullKRaft` we have right now before removing ZooKeeper. It could be named something else, or if we decide to go with more automation (something I am not sure right now tbh) it could just be removed.